lib/AstOfLlbc.ml

```diff
@@ -628,39 +896,77 @@ let fn_ptr_is_opaque env (fn_ptr: C.fn_ptr) =
+  L.log "Calls" "%s--> %d type_args, %d const_generics, %d trait_refs" depth
     (List.length type_args) (List.length const_generic_args) (List.length trait_refs);
 
   let type_args, const_generic_args, trait_refs =
     match func with
-    | TraitMethod ({ generics = { types; const_generics; trait_refs; _ }; _ }, _, _) ->
-        types @ type_args, const_generics @ const_generic_args, trait_refs @ trait_refs
+    | TraitMethod ({ generics = { types; const_generics; trait_refs = trait_refs'; _ }; _ }, _, _) ->
+        (* TODO: why is the order fumbled here? *)
+        types @ type_args, const_generics @ const_generic_args, trait_refs @ trait_refs'
     | _ ->
         type_args, const_generic_args, trait_refs
   in
 
-  L.log "Calls" "--> %d type_args, %d const_generics, %d trait_refs"
+  L.log "Calls" "%s--> %d type_args, %d const_generics, %d trait_refs" depth
     (List.length type_args) (List.length const_generic_args) (List.length trait_refs);
-  L.log "Calls" "--> trait_refs: %s\n"
+  L.log "Calls" "%s--> trait_refs: %s\n" depth
     (String.concat " ++ " (List.map (Charon.PrintTypes.trait_ref_to_string env.format_env) trait_refs));
-  L.log "Calls" "--> pattern: %s" (string_of_fn_ptr env fn_ptr);
-  L.log "Calls" "--> type_args: %s" (String.concat ", " (List.map (Charon.PrintTypes.ty_to_string env.format_env) type_args));
+  L.log "Calls" "%s--> pattern: %s" depth (string_of_fn_ptr env fn_ptr);
+  L.log "Calls" "%s--> type_args: %s" depth (String.concat ", " (List.map (Charon.PrintTypes.ty_to_string env.format_env) type_args));
 
   let type_args = List.map (typ_of_ty env) type_args in
   let const_generic_args = List.map (expression_of_const_generic env) const_generic_args in
-  let { name; n_type_args = n_type_params; arg_types = inputs; ret_type = output; cg_types = cg_inputs; is_assumed } =
-    lookup_fun env fn_ptr
+  let f, { n_type_args = n_type_params; arg_types = inputs; ret_type = output; cg_types = cg_inputs; is_known_builtin } =
+    lookup_fun env depth fn_ptr
+  in
+
+  (* Handling trait implementations for generic trait bounds in the callee. *)
+  let fn_ptrs =
+    if is_known_builtin then
+      (* If this is a known builtin implementation, we do not materialize trait methods, on the
+         basis that this is likely something from the standard library that exercises more features
+         that we can support, and that since we hand-write it, we don't need this level of precision
+         anyhow. *)
+      []
+    else
+      List.concat_map (fun (trait_ref: C.trait_ref) ->
+        let name = string_of_name env (env.get_nth_trait_decl trait_ref.trait_decl_ref.trait_decl_id).name in
+
+        if List.mem name blocklisted_trait_decls then
+          []
+
+        else
+          let trait_impl: C.trait_impl =
+            match trait_ref.trait_id with
+            | TraitImpl impl_id -> env.get_nth_trait_impl impl_id
+            | _ -> failwith ("impossible: " ^ C.show_trait_ref trait_ref)
```

Comment:
Why is this supposed to be impossible?

The following triggers a case where the `trait_impl` is a `Clause`:
```rust
trait Marker {}

trait Trait {
    fn f();
}

impl<T: Marker> Trait for T {
    fn f() { }
}

fn bug<T: Marker>(x: T) {
    T::f()
}
```

Here, in function `bug`, the call `T::f` is actually `(impl_0::<Clause 0>)::f`, where `Clause 0` refers to the generic bound `<T: Marker>` on function `bug`.
The concrete impl block `impl<T: Marker> Trait for T` has one generic `T` which requires an impl for `Marker`.

If I understand correctly, we are here trying to construct a `trait_impl` for every bound that occurs both on the function itself but also on the impl block?

Looking at the type `C.trait_impl`, I wonder if that type can represent such local bounds (here we have nothing but a bound variable).

Reply:
Actually, this example also triggers the same "impossible" error:
```rust
trait Trait {}

fn f<T: Trait>() {}
fn g<T: Trait>() {
    f::<T>()
}
```

Reply:
This has since then been fixed with support for trait bounds implemented as dictionary-passing style. Thanks for the minimal repro.